### PR TITLE
Add list and delete commands in policy recommendation CLI

### LINKED ISSUE #49: List and delete policy recommendation jobs
#### Add a list command to show all policy recommendation jobs, e.g:
```bash
> theia pr list
CreateTime          CompleteTime        ID                                   Status
2022-06-17 18:33:15 N/A                 2cf13427-cbe5-454c-b9d3-e1124af7baa2 RUNNING
2022-06-17 18:06:56 2022-06-17 18:08:37 69e2e543-60e9-4d45-97a1-d56337966579 COMPLETED
2022-06-16 23:41:43 2022-06-16 23:43:15 a65daf22-8e7e-4479-9f4e-edc1d99716ff COMPLETED
N/A                 2022-06-13 22:19:17 749ecc41-bf5e-4d08-88ef-fb66b60bf1fb COMPLETED
N/A                 2022-06-15 21:41:16 1e7ffc6d-2321-422d-b982-0ffca2d7987f COMPLETED
```
We will fetch and display all `sparkapplication` from the K8s API server first, then we will check the recommendation result table of ClickHouse DB for additional completed jobs that are not inside the K8s API server (probably they are deleted by users manually).
`CompleteTime` of uncompleted jobs will show as `N/A`, `CreateTime` of jobs fetching from the result table of ClickHouse DB will show as `N/A` since we didn't save that info in DB.

#### Add a delete command to let users delete a policy recommendation job given ID. 
If the status of this job is not completed(like running, falied, etc), the `sparkapplication` behind this job will be deleted. Otherwise, both `sparkapplication` behind this job and the recommendation result in database will be deleted. e.g:
```bash
> theia pr delete 1e7ffc6d-2321-422d-b982-0ffca2d7987f
Successfully deleted policy recommendation job with ID 1e7ffc6d-2321-422d-b982-0ffca2d7987f
```

#### Open question:
1. Do we need to add a `status` column in the recommendations table?
Currently, statuses of policy recommendation jobs are obtained from the k8s API server. Only completed jobs will write results into the recommendations table.
May need to think about how to sync status between the k8s API server and database. (Should be handled in the middle layer application later)
For now, let's not add more columns to the recommendation result table.

2. Do we have other columns that would like to be added to the list command result?
job parameters, flows number, etc. 
(For now we don't plan to show job parameters in list command since they are too much to display in a table, users could see them by describe the driver pod)

3. Check the failure reason for failed jobs and write them into database
Could check the debug APIs of spark operator first.
There is a `Error Message` field in failed `sparkapplication`, we could display it in the `status` command of cli.
Writing it into database should be handled in the middle layer application later.

### PR DESCRIPTION
**Pull request by dreamtalen**
Fix #49 

Signed-off-by: Yongming Ding <dyongming@vmware.com>

### CODE REVIEW

File: docs/networkpolicy-recommendation.md [hunk 1]
```diff
@@ -134,3 +140,26 @@ policies to a YAML file and apply it using `kubectl`:
 theia policy-recommendation retrieve e998433e-accb-4888-9fc8-06563f073e86 -f recommended_policies.yml
 kubectl apply -f recommended_policies.yml
 ```
+
+### List all policy recommendation jobs
+
+The `theia policy-recommendation list` command lists all policy recommendation
+jobs. `CreateTime`, `CompleteTime`, `ID` and `Status` of each policy
+recommendation job will be displayed in the form of a table. For example:
```
Comment (annakhm) [inline docs/networkpolicy-recommendation.md:148]: suggestion: in table format

File: pkg/theia/commands/policy_recommendation_delete.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+		return idMap, err
+	}
+	for _, completedPolicyRecommendation := range completedPolicyRecommendationList {
+		idMap[completedPolicyRecommendation.id] = true
+	}
+	return idMap, nil
+}
+
+func deletePolicyRecommendationResult(clientset kubernetes.Interface, kubeconfig string, endpoint string, useClusterIP bool, recoID string) (err error) {
+	if endpoint == "" {
```
Comment (annakhm) [inline pkg/theia/commands/policy_recommendation_delete.go:128]: there seem to be a lot of code duplication with `getPolicyRecommendationResult`, could the common code be reused?

File: pkg/theia/commands/policy_recommendation_list.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+			}
+		}
+
+		TableOutput(sparkApplicationTable)
+		return nil
+	},
+}
+
+func getCompletedPolicyRecommendationList(clientset kubernetes.Interface, kubeconfig string, endpoint string, useClusterIP bool) (completedPolicyRecommendationList []policyRecommendationRow, err error) {
+	if endpoint == "" {
```
Comment (annakhm) [inline pkg/theia/commands/policy_recommendation_list.go:125]: ditto

File: pkg/theia/commands/policy_recommendation_delete.go (hunk 2 above)
Comment (dreamtalen) [reply pkg/theia/commands/policy_recommendation_delete.go]: Sure, I refactored the code setting up the ClickHouse connection among `list`, `retrieve`, and `delete` commands.

File: pkg/theia/commands/policy_recommendation_list.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+	rows, err := connect.Query(query)
+	if err != nil {
+		return completedPolicyRecommendationList, fmt.Errorf("failed to get recommendation jobs: %v", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var row policyRecommendationRow
+		err := rows.Scan(&row.timeComplete, &row.id)
+		if err != nil {
+			return completedPolicyRecommendationList, fmt.Errorf("err when scaning recommendations row %v", err)
```
Comment (yanjunz97) [inline pkg/theia/commands/policy_recommendation_list.go:142]: err when scaning recommendations row %v -> error when scanning recommendations row: %v?

File: pkg/theia/commands/policy_recommendation_list.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+		"clickhouse-endpoint",
+		"",
+		"The ClickHouse service endpoint.",
+	)
+	policyRecommendationListCmd.Flags().Bool(
+		"use-cluster-ip",
+		false,
+		`Enable this option will use Service ClusterIP instead of port forwarding when connecting to the ClickHouse service.
+It can only be used when running theia in cluster.`,
+	)
```
Comment (yanjunz97) [inline pkg/theia/commands/policy_recommendation_list.go:161]: Not sure whether it would be better to put this into global flags, as I think up to now all commands requires this flag.

File: docs/networkpolicy-recommendation.md [hunk 6]
```diff
@@ -134,3 +140,26 @@ policies to a YAML file and apply it using `kubectl`:
...
+> theia policy-recommendation list
+CreateTime          CompleteTime        ID                                   Status
+2022-06-17 18:33:15 N/A                 2cf13427-cbe5-454c-b9d3-e1124af7baa2 RUNNING
+2022-06-17 18:06:56 2022-06-17 18:08:37 e998433e-accb-4888-9fc8-06563f073e86 COMPLETED
+```
+
+### Delete a policy recommendation job
+
+The `theia policy-recommendation delete` command is used to delete a policy
+recommendation job. To delete the policy recommendation job created above, run:
```
Comment (yanjunz97) [inline docs/networkpolicy-recommendation.md:160]: I notice there is a rare case that if users try to delete a job before the driver pod is created, the job may still keep running till completion. And users will receive a message saying the job is successfully deleted. Not sure if we should deal with this case by returning another message or doc it here.

**Comment by yanjunz97** [comment]
Not introduced by this PR, I just recall previously there is a discussion on replacing the empty state string when we get a [NewState](https://pkg.go.dev/github.com/GoogleCloudPlatform/spark-on-k8s-operator/pkg/apis/sparkoperator.k8s.io/v1beta2#ApplicationStateType) by running `theia pr status xxx`. Maybe we can fix it here or in a new PR?

File: pkg/theia/commands/policy_recommendation_list.go (hunk 5 above)
Comment (yuntanghsu) [reply pkg/theia/commands/policy_recommendation_list.go:161]: I think both `clickhouse-endpoint` and `use-cluster-ip` might be added to glabal flags?

Comment (dreamtalen) [reply pkg/theia/commands/policy_recommendation_list.go:161]: Hi Yanjun and Yuntang, yes there are some duplicates here, but `status` command uses `use-cluster-ip` only and is used for `Spark Monitoring Service` instead of `ClickHouse Service` in other commands.
Ultimately, I think the connection to `Spark Monitoring Service` and `ClickHouse Service` should be handled by the middle layer app theia manager. We could remove these flags then imo.

File: test/e2e/policyrecommendation_test.go [hunk 7]
```diff
@@ -97,24 +107,53 @@ func TestPolicyRecommendation(t *testing.T) {
...
-	jobId := stdoutSlice[len(stdoutSlice)-1]
-	stdout, err = getJobStatus(t, data, jobId)
+	stdout, err := getJobStatus(t, data, jobId)
 	require.NoError(t, err)
 	assert := assert.New(t)
 	assert.Containsf(stdout, "Status of this policy recommendation job is", "stdout: %s", stdout)
 }
 
+// Example output:
+// CreateTime          ID                                   Status
```
[the remark this answers is not in the document]
Comment (dreamtalen) [reply test/e2e/policyrecommendation_test.go:127]: Sure, I forget to change here, thanks.

File: docs/networkpolicy-recommendation.md (hunk 6 above)
Comment (dreamtalen) [reply docs/networkpolicy-recommendation.md:160]: Thanks Yanjun, I have seen this corner case too during e2e test. In that case, the `sparkapplication` cr has been deleted successfully but the deriver/exec pods will still be created later, maybe I should create an issue in the spark operator repo to hear their solutions.

**Comment by dreamtalen** [comment]
> Not introduced by this PR, I just recall previously there is a discussion on replacing the empty state string when we get a [NewState](https://pkg.go.dev/github.com/GoogleCloudPlatform/spark-on-k8s-operator/pkg/apis/sparkoperator.k8s.io/v1beta2#ApplicationStateType) by running `theia pr status xxx`. Maybe we can fix it here or in a new PR?

No problem, I will take care of it.

File: pkg/theia/commands/policy_recommendation_list.go (hunk 5 above)
Comment (yuntanghsu) [reply pkg/theia/commands/policy_recommendation_list.go:161]: I think for ClickHouse service, we also need these two flags in order to setup clickhouse connection? Currently, I add these two flags when using `theia clickhouse status` in PR #59

Comment (dreamtalen) [reply pkg/theia/commands/policy_recommendation_list.go:161]: For our `theia pr status` command, we're fetching status from k8s api server and spark monitoring service, that's why it doesn't need to connect with clickhouse for now.

But I also realize that we could improve the `status` command to let it can check completed jobs in clickhouse only. In that case, all commands needs connection with clickhouse, I plan to create a separate PR for this improvement of `status` command first.

File: pkg/theia/commands/utils.go [hunk 8]
```diff
@@ -156,3 +161,103 @@ func ResolveKubeConfig(cmd *cobra.Command) (string, error) {
...
+	connRetryInterval := 1 * time.Second
+	connTimeout := 10 * time.Second
+
+	// Connect to ClickHouse in a loop
+	if err := wait.PollImmediate(connRetryInterval, connTimeout, func() (bool, error) {
+		// Open the database and ping it
+		var err error
+		connect, err = sql.Open("clickhouse", url)
+		if err != nil {
+			connErr = fmt.Errorf("failed to ping ClickHouse: %v", err)
```
Comment (yuntanghsu) [inline pkg/theia/commands/utils.go:208]:
```suggestion
			connErr = fmt.Errorf("failed to open ClickHouse: %v", err)
```
?

File: docs/networkpolicy-recommendation.md [hunk 9]
```diff
@@ -134,3 +140,26 @@ policies to a YAML file and apply it using `kubectl`:
 theia policy-recommendation retrieve e998433e-accb-4888-9fc8-06563f073e86 -f recommended_policies.yml
 kubectl apply -f recommended_policies.yml
 ```
+
+### List all policy recommendation jobs
+
+The `theia policy-recommendation list` command lists all policy recommendation
```
Comment (ziyouw) [inline docs/networkpolicy-recommendation.md:146]: All undeleted jobs?

File: docs/networkpolicy-recommendation.md (hunk 6 above)
Comment (ziyouw) [inline docs/networkpolicy-recommendation.md:160]: Since the deletion cannot be undone, maybe we need to warn this.

Comment (dreamtalen) [reply docs/networkpolicy-recommendation.md]: Sure, added.

**Review by ziyouw — APPROVED** [review]
LGTM, thanks!

File: pkg/theia/commands/policy_recommendation_list.go [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+			return fmt.Errorf("couldn't create k8s client using given kubeconfig, %v", err)
+		}
+		endpoint, err := cmd.Flags().GetString("clickhouse-endpoint")
+		if err != nil {
+			return err
+		}
+		if endpoint != "" {
+			_, err := url.ParseRequestURI(endpoint)
+			if err != nil {
+				return fmt.Errorf("failed to decode input endpoint %s into a url, err: %v", endpoint, err)
```
[the remark this answers is not in the document]
Comment (dreamtalen) [reply pkg/theia/commands/policy_recommendation_list.go:61]: Sure, add two utility functions for parsing ID and endpoint.

File: pkg/theia/commands/policy_recommendation_list.go [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+		}
+
+		completedPolicyRecommendationList, err := getCompletedPolicyRecommendationList(clientset, kubeconfig, endpoint, useClusterIP)
+
+		if err != nil {
+			return err
+		}
+
+		sparkApplicationTable := [][]string{
+			{"CreateTime", "CompleteTime", "ID", "Status"},
```
[the remark this answers is not in the document]
Comment (dreamtalen) [reply pkg/theia/commands/policy_recommendation_list.go:91]: Sure, to be consistent, I'm changing to use `CreationTime` and `CompletionTime`.

File: pkg/theia/commands/policy_recommendation_delete.go [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+
+func deletePolicyRecommendationResult(clientset kubernetes.Interface, kubeconfig string, endpoint string, useClusterIP bool, recoID string) (err error) {
+	connect, portForward, err := setupClickHouseConnection(clientset, kubeconfig, endpoint, useClusterIP)
+	if portForward != nil {
+		defer portForward.Stop()
+	}
+	if err != nil {
+		return err
+	}
+	query := "ALTER TABLE recommendations DELETE WHERE id = (?);"
```
[the remark this answers is not in the document]
Comment (dreamtalen) [reply pkg/theia/commands/policy_recommendation_delete.go:133]: Yes, Clickhouse doesn't have Update/Delete commands like Mysql database, reference doc: https://clickhouse.com/docs/en/sql-reference/statements/alter/delete/

**Review by heanlan — APPROVED** [review]
